designs/auto-certs-renewal.md

```diff
@@ -0,0 +1,121 @@
+
+### Worker node certificates
+
+1. just ca.crt in /etc/kubernetes/pki, which should be generated after `kubeadm join`
+
+## Goal
+
+1. Auto renew all the above mentioned certificates except etcd CA and kube CA, which last for 10 years.
+
+### Observation
+
+1. We can renew all etcd’s certs by simply **recreating the external etcd machine**. If doing manually, we just need to delete the auto generated certs and run `sudo etcdadm join phase certificates`
+2. We can renew all cp’s certs besides apiserver-etcd-client by **recreating the cp machine.** As we can see from the following diagram that the apiserver-etcd-client is generated by Etcdadm Controller, so this cert cannot be renewed by just recreating the cp machine.
+3. No need to do auto renew certs in worker nodes.
+
+## Solution
+
+Etcd can be deployed in [two topologies.](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/) Certs renew mechanism is slightly different in those two scenarios.
+
+### Stacked etcd Nodes
+
+In Stacked etcd nodes topology, etcd is running inside the cp node. 
+
+We can renew all certificates by just redeploying cp nodes
+
+1. Expose [rolloutBefore.certificatesExpiryDays](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp.html) field from KubeadmControlPlane in EKSA cluster spec. Add pass the field to KubeadmControlPlane
+
+### External etcd Nodes
+
+In external etcd nodes topology, etcd is running in a separate machine besides cp machine. In addition to roll out cp nodes, we also need to roll out etcd nodes, with the following changes on top of above:
+
+1. Add rolloutBefore.certificatesExpiryDays field to [EtcdadmCluster](https://github.com/aws/etcdadm-controller/blob/main/api/v1beta1/etcdadmcluster_types.go#L38), exposing and passing it from EKSA. The business logic of this field and reconciler should be:
+    1. apply machine annotation `machine.cluster.x-k8s.io/certificates-expiry` during etcd node creation.
+    2. check if roll out is needed. Refer https://github.com/kubernetes-sigs/cluster-api/blob/main/controlplane/kubeadm/internal/controllers/controller.go#L380.
+    3. if roll out is needed for any etcd node, upgrade all etcd nodes.
+
+However, not all certificates in cp node would be renewed even both etcd and cp nodes have changed, since apiserver-etcd-client key pair is created by etcdadm controller (refer the sequence diagram) and it always stays the same. We need to fix this bug by changing [etcdadm controller’s code](https://github.com/aws/etcdadm-controller/blob/main/controllers/certs.go#L81), it should update apiserver-etcd-client secret constantly in a short interval, such as every 24 hours. If we choose just update the secret during etcd nodes rolling out, then we have to invent mechanism to make sure the etcd nodes rolling out happens before cp nodes rolling out when it comes to renew certificates.
```

Comment:
This is still necessary. It is a bug existing today: a user would have apiserer-etcd-client expired even if he did the rollout for etcd and cp everyday.

The proposed solution by rollout etcd only would only make sure that cp doesn't have an extra rollout

Reply:
OK, help me walk through this pls, I might be missing something.
The etcd rollout to renew certs will always happen before the CP rollout. In fact, it's etcd the one that will cause the CO to rollout.

Then, why can't we just renew the apiserver-etcd-client secret just before the etcd rollout? That way it will be fresh when the CP nodes get rolled out.

Reply:
If we just renew apiserver-etcd-client cert before etcd rollout, the apiserver-etcd-client cert of control plane will be renewed. However, the bug still exists: as a user, when I rollout cp nodes, I want all certs in the cp nodes get renewed.

Reply:
why is it a bug? There is no contract saying that the client cert always has to be fresh.

> However, the bug still exists: as a user, when I rollout cp nodes, I want all certs in the cp nodes get renewed.

Where is this requirement coming from?

Reply:
I would say it's an implicit requirement based on common sense.

If a user upgrades his cluster everyday (not enabling auto certs renew), and he still get this cert expired in 1 year, he would have a great chance calling this behaviour "bug".

Reply:
I think we need to talk through the problem and solution some more. Are we able to schedule a call?